### PR TITLE
Add Branch Coverage data for ProfData coverage files

### PR DESCRIPTION
**Pull request by hborawski**
Using the [lcov json data](https://llvm.org/doxygen/structllvm_1_1coverage_1_1CoverageSegment.html) segments, populate `branch_coverage_data` for profdata coverage files

This allows the cobertura report to correctly calculate branch coverage

Also added branch coverage to the HTML output by using the segments to determine column ranges that were not hit, so null coalesecing operations, or other single line branches can be determined as missed coverage.

Please let me know if there are any updates required for this PR, it's been quite some time since using ruby, so I'm sure the code can be improved.

I made sure the existing tests pass, and i added some small profdata tests to check that the branch data populates as expected.

I've also tested this locally against a real project and compared output to Xcode, to see how it works on a larger codebase.

Example of a covered line with an uncovered branch:
Screen Shot 2021-02-23 at 2 52 52 PM

### CODE REVIEW

File: assets/slather.css [hunk 1]
```diff
@@ -74,6 +74,7 @@ table.source_code {
 table.source_code td {
   padding-bottom: 0.3em;
 }
+code.missed {background-color: rgba(255, 73, 76, 0.3);}
```
Comment (ksuther) [inline assets/slather.css:77]: Code style nitpick: Add space after and before braces to match the other lines

File: lib/slather/coverage_service/html_output.rb [hunk 2]
```diff
@@ -133,6 +144,7 @@ def create_html_from_file(coverage_file)
         filepath = coverage_file.source_file_pathname_relative_to_repo_root
         filename = File.basename(filepath)
         percentage = coverage_file.percentage_lines_tested
+        branchPercentage = coverage_file.rate_branches_tested * 100
```
Comment (ksuther) [inline lib/slather/coverage_service/html_output.rb:147]: Code style nitpick: Use snake_case (branch_percentage)

File: lib/slather/coverage_service/html_output.rb [hunk 3]
```diff
@@ -157,8 +172,9 @@ def create_html_from_file(coverage_file)
 
           cov.table(:class => "source_code") {
             cleaned_gcov_lines.each do |line|
-
               line_number = coverage_file.line_number_in_line(line)
+              missedRegions = coverage_file.branch_region_data[line_number]
```
Comment (ksuther) [inline lib/slather/coverage_service/html_output.rb:176]: Code style nitpick: Use snake_case (missed_regions)

File: lib/slather/coverage_service/html_output.rb [hunk 4]
```diff
@@ -171,7 +187,30 @@ def create_html_from_file(coverage_file)
                     cov.td(line, :class => classes[idx])
                   else
                     cov.td(:class => classes[idx]) {
-                      cov.pre { cov.code(line, :class => "objc") }
+                      cov.pre {
+                        # If the line has coverage and missed regions, split up
+                        # the line to show regions that weren't covered
+                        if missedRegions != nil && hits != nil && hits > 0
+                          regions = missedRegions.map do |region|
+                            regionStart, regionLength = region
```
Comment (ksuther) [inline lib/slather/coverage_service/html_output.rb:195]: Code style nitpick: Use snake_case (region_start, region_length)

File: lib/slather/coverage_service/html_output.rb [hunk 5]
```diff
@@ -171,7 +187,30 @@ def create_html_from_file(coverage_file)
...
+                            if regionLength != nil
+                              line[regionStart, regionLength]
+                            else
+                              line[regionStart, line.length - regionStart]
+                            end
+                          end
+                          currentLine = line
```
Comment (ksuther) [inline lib/slather/coverage_service/html_output.rb:202]: Code style nitpick: Use snake_case (current_line)

File: lib/slather/profdata_coverage_file.rb [hunk 6]
```diff
@@ -188,7 +188,47 @@ def coverage_for_line(line, line_numbers_first = self.line_numbers_first)
...
+        end
+
+        branch_coverage_data
+      end
+    end
+
+    def branch_region_data
+      @branch_region_data ||= begin
+        branch_region_data = Hash.new
+        regionStart = nil
```
Comment (ksuther) [inline lib/slather/profdata_coverage_file.rb:210]: Code style nitpick: Use snake_case (region_start)

File: lib/slather/profdata_coverage_file.rb [hunk 7]
```diff
@@ -188,7 +188,47 @@ def coverage_for_line(line, line_numbers_first = self.line_numbers_first)
...
+        currentLine = 0
+        @segments ||= []
+        @segments.each do |segment|
+          line, col, hits, hasCount, *rest = segment
```
Comment (ksuther) [inline lib/slather/profdata_coverage_file.rb:214]: Code style nitpick: Use snake_case (has_count)

File: lib/slather/profdata_coverage_file.rb [hunk 8]
```diff
@@ -188,7 +188,47 @@ def coverage_for_line(line, line_numbers_first = self.line_numbers_first)
...
+
+        branch_coverage_data
+      end
+    end
+
+    def branch_region_data
+      @branch_region_data ||= begin
+        branch_region_data = Hash.new
+        regionStart = nil
+        currentLine = 0
```
Comment (ksuther) [inline lib/slather/profdata_coverage_file.rb:211]: Code style nitpick: Use snake_case (current_line)

File: lib/slather/project.rb [hunk 9]
```diff
@@ -162,13 +162,24 @@ def create_coverage_files_for_binary(binary_path, pathnames_per_binary)
     end
     private :create_coverage_files_for_binary
 
-    def create_coverage_files(binary_path, pathnames)
+    def create_coverage_files(binary_path, pathObjects)
```
Comment (ksuther) [inline lib/slather/project.rb:165]: Code style nitpick: Use snake_case (path_objects)

File: lib/slather/project.rb [hunk 10]
```diff
@@ -162,13 +162,24 @@ def create_coverage_files_for_binary(binary_path, pathnames_per_binary)
...
       line_numbers_first = Gem::Version.new(self.llvm_version) >= Gem::Version.new('8.1.0')
+      # get just file names from the path objects
+      pathnames = pathObjects.map { |pathObj| pathObj["filename"] }.compact
```
Comment (ksuther) [inline lib/slather/project.rb:168]: Code style nitpick: Use snake_case (path_obj)

File: lib/slather/project.rb [hunk 11]
```diff
@@ -162,13 +162,24 @@ def create_coverage_files_for_binary(binary_path, pathnames_per_binary)
...
+      # Map of path name => segment array
+      pathsToSegments = pathObjects.reduce(Hash.new) do |hash, pathObj|
```
Comment (ksuther) [inline lib/slather/project.rb:170]: Code style nitpick: Use snake_case (paths_to_segments)

**Review by ksuther — COMMENTED** [review]
Thanks for the PR! I made some incredibly nitpicky suggestions just to match the existing code style.

I'm by no means a Ruby pro either, so if anyone else has any suggestions feel free to chime in. Otherwise if it's working on your code base and the tests are passing I'm happy to merge this in and cut a release.

**Comment by hborawski** [comment]
Updated the code to use snake case for all the variables, didn't even realize the rest of the codebase was snake 🤦

**Comment by ksuther** [comment]
Merged, thanks for the contribution! I'll make a release to include this.

**Comment by hborawski** [comment]
Happy to help, and excited to get all my projects updated with this!

**Pull request merged**